packages/js/product-editor/src/components/details-name-block/edit.tsx

```diff
@@ -2,36 +2,175 @@
+	__experimentalInputControl as InputControl,
+} from '@wordpress/components';
 // eslint-disable-next-line @typescript-eslint/ban-ts-comment
 // @ts-ignore No types for this exist yet.
 // eslint-disable-next-line @woocommerce/dependency-group
-import { useEntityProp } from '@wordpress/core-data';
+import { useEntityProp, useEntityId } from '@wordpress/core-data';
+
+/**
+ * Internal dependencies
+ */
+import { AUTO_DRAFT_NAME } from '../../utils';
+import { EditProductLinkModal } from '../edit-product-link-modal';
+import { useValidation } from '../../hooks/use-validation';
+import { useProductHelper } from '../../hooks/use-product-helper';
 
 export function Edit() {
 	const blockProps = useBlockProps();
-	const [ name, setName ] = useEntityProp( 'postType', 'product', 'name' );
+
+	const { updateProductWithStatus } = useProductHelper();
+
+	const [ showProductLinkEditModal, setShowProductLinkEditModal ] =
+		useState( false );
+
+	const { editEntityRecord } = useDispatch( 'core' );
+	const productId = useEntityId( 'postType', 'product' );
+	const product: Product = useSelect( ( select ) =>
+		select( 'core' ).getEditedEntityRecord(
+			'postType',
+			'product',
+			productId
+		)
+	);
+
+	const [ name, setName ] = useEntityProp< string >(
+		'postType',
+		'product',
+		'name'
+	);
+
+	const { permalinkPrefix, permalinkSuffix } = useSelect(
+		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+		// @ts-ignore
+		( select: WCDataSelector ) => {
+			const { getPermalinkParts } = select( PRODUCTS_STORE_NAME );
+			if ( productId ) {
+				const parts = getPermalinkParts( productId );
+				return {
+					permalinkPrefix: parts?.prefix,
+					permalinkSuffix: parts?.suffix,
+				};
+			}
+			return {};
+		}
+	);
+
+	const nameIsValid = useValidation(
+		'product/name',
+		() => Boolean( name ) && name !== AUTO_DRAFT_NAME
+	);
 
 	return (
-		<div { ...blockProps }>
-			<TextControl
-				label={ createInterpolateElement(
-					__( 'Name <required />', 'woocommerce' ),
-					{
-						required: (
-							<span className="woocommerce-product-form__optional-input">
-								{ __( '(required)', 'woocommerce' ) }
-							</span>
-						),
-					}
+		<>
+			<div { ...blockProps }>
+				<BaseControl
+					id={ 'product_name' }
+					label={ createInterpolateElement(
+						__( 'Name <required />', 'woocommerce' ),
+						{
+							required: (
+								<span className="woocommerce-product-form__required-input">
+									{ __( '*', 'woocommerce' ) }
+								</span>
+							),
+						}
+					) }
+				>
+					<InputControl
+						name={ 'woocommerce-product-name' }
+						placeholder={ __(
+							'e.g. 12 oz Coffee Mug',
+							'woocommerce'
+						) }
+						onChange={ setName }
+						value={ name || '' }
+					/>
+				</BaseControl>
+				{ productId &&
+					nameIsValid &&
+					product.status === 'publish' &&
+					permalinkPrefix && (
+						<span className="woocommerce-product-form__secondary-text product-details-section__product-link">
```

Comment:
Was there a specific reason for integrating the product link stuff within the details name block versus creating a separate block for this?
I am not entirely opposed to this, as both the product name and product link are relatively intertwined and are probably always shown next to each other, but curious if you had additional reasons? Maybe styling also?